content/learning-paths/cross-platform/simd-on-rust/simd-on-rust-part3.md

```diff
@@ -0,0 +1,218 @@
+ 8f0:   6d400c00        ldp     d0, d3, [x0]
+ 8f4:   6d411001        ldp     d1, d4, [x0, #16]
+ 8f8:   0e432802        trn1    v2.4h, v0.4h, v3.4h
+ 8fc:   0e436800        trn2    v0.4h, v0.4h, v3.4h
+ 900:   0e442823        trn1    v3.4h, v1.4h, v4.4h
+ 904:   0e446821        trn2    v1.4h, v1.4h, v4.4h
+ 908:   0e833844        zip1    v4.2s, v2.2s, v3.2s
+ 90c:   0e837842        zip2    v2.2s, v2.2s, v3.2s
+ 910:   0e813803        zip1    v3.2s, v0.2s, v1.2s
+ 914:   0e817800        zip2    v0.2s, v0.2s, v1.2s
+ 918:   6d000c04        stp     d4, d3, [x0]
+ 91c:   6d010002        stp     d2, d0, [x0, #16]
+ 920:   d65f03c0        ret
+```
+
+Pretty straightforward, as expected the relevant instructions are executed in sequence without any surprises.
+
+Let's look at the equivalent Rust implementation:
+
+```Rust
+const N : usize = 4;
+
+fn main() {
+    let mut a: [i16; N*N] = [0; N*N];
+
+    init_vec(&mut a);
+    print_vec(&a);
+    transpose_s16_4x4(&mut a);
+    print_vec(&a);
+}
+
+fn init_vec(a: &mut [i16; N*N]) -> () {
+    for i in 0..N {
+        for j in 0..N {
+            a[i*N + j] = (4*i + j + 1) as i16;
+        }
+    }
+}
+
+fn print_vec(a: &[i16; N*N]) -> () {
+    println!("A[] =");
+    for i in 0..N {
+        for j in 0..N {
+            print!("{:02x} ", a[i*N + j]);
+        }
+        println!();
+    }
+}
+
+#[inline(never)]
+fn transpose_s16_4x4(a: &mut [i16]) -> () {
+    #[cfg(target_arch = "aarch64")]
+    {
+        use std::arch::is_aarch64_feature_detected;
+        if is_aarch64_feature_detected!("neon") {
+            return unsafe { transpose_s16_4x4_asimd(a) };
+        }
+    }
+    // Scalar implementation should be included here as fallback
+}
+
+#[cfg(target_arch = "aarch64")]
+#[target_feature(enable = "neon")]
+unsafe fn transpose_s16_4x4_asimd(a: &mut [i16]) -> () {
+    use std::arch::aarch64::*;
+    let va: [int16x4_t; N] = [ vld1_s16(&a[0]), vld1_s16(&a[4]), vld1_s16(&a[8]), vld1_s16(&a[12]) ];
+
+    let b0 : int16x4x2_t = vtrn_s16(va[0], va[1]);
+    let b1 : int16x4x2_t = vtrn_s16(va[2], va[3]);
+
+    let c0 : int32x2x2_t = vtrn_s32(vreinterpret_s32_s16(b0.0),
+                                    vreinterpret_s32_s16(b1.0));
+    let c1 : int32x2x2_t = vtrn_s32(vreinterpret_s32_s16(b0.1),
+                                    vreinterpret_s32_s16(b1.1));
+
+    vst1_s16(&mut a[0], vreinterpret_s16_s32(c0.0));
+    vst1_s16(&mut a[4], vreinterpret_s16_s32(c1.0));
+    vst1_s16(&mut a[8], vreinterpret_s16_s32(c0.1));
+    vst1_s16(&mut a[12], vreinterpret_s16_s32(c1.1));
+}
+```
+
+You will note two important differences here:
+
+* initialization of the `int16x4_t va` array at the start of the `transpose_s16_4x4_asimd` function. In C you can just declare the variable `va` and the compiler may not complain much. Even with the relevant compile flags, it will understand that the SIMD variable will be immediately initialized by the respective load instructions `vld1q_s16`.
```

Comment:
It didn't in my example, even though it doesn't appear in the generated assembly, I still had to zero-initialize va, the compiler complained otherwise. I don't know if it's something that's missing from the compiler, but I mention it here as others will probably check that behaviour as well.